source/common/upstream/health_checker_impl.cc

```diff
@@ -358,21 +360,23 @@ void TcpHealthCheckerImpl::TcpActiveHealthCheckSession::onData(Buffer::Instance&
     data.drain(data.length());
     handleSuccess(false);
     if (!parent_.reuse_connection_) {
-      client_->close(Network::ConnectionCloseType::NoFlush);
+      Network::ClientConnectionPtr clientToDestroy(std::move(client_));
+      clientToDestroy->close(Network::ConnectionCloseType::NoFlush);
+      parent_.dispatcher_.deferredDelete(std::move(clientToDestroy));
     }
   } else {
     host_->setActiveHealthFailureType(Host::ActiveHealthFailureType::UNHEALTHY);
   }
 }
 
 void TcpHealthCheckerImpl::TcpActiveHealthCheckSession::onEvent(Network::ConnectionEvent event) {
-  if (event == Network::ConnectionEvent::RemoteClose) {
+  // If !client_, then we are already handling a failure/teardown
```

Comment:
Key is this.  I'm using `!client_` as a signal that this is an internally-generated `RemoteClose` or `LocalClose`.  I need some way to distinguish so I know that I need to call `handleFailure()`.

(Other ways to distinguish internally-generated from externally-generated would work too)

Reply:
The other way to do this would be to do what the other health checkers are doing which is effectively the `expect_reset` boolean. I'm fine either way.

Reply:
I updated to follow the `expect_reset_` boolean pattern for consistency with the other health checkers, thanks for the suggestion.